Public/Src/Cache/ContentStore/Distributed/NuCache/BinManager.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using BuildXL.Cache.ContentStore.Hashing;
+using BuildXL.Cache.ContentStore.Utils;
+
+namespace BuildXL.Cache.ContentStore.Distributed.NuCache
+{
+    internal class BinManager
+    {
+        private const int MaxBins = 64 * 1024;
+
+        private readonly MachineLocation[] _bins;
+        private readonly int _locationsPerBin;
+        private readonly int _entriesPerMachine;
+
+        private MachineLocation[][] _machinesForBins;
+
+        private readonly IContentHasher _contentHasher = ContentHashers.Get(HashType.MD5);
+
+        /// <nodoc />
+        public BinManager(int machinesPerBin, int entriesPerMachine, int amountOfBins)
+        {
+            Contract.Assert(entriesPerMachine <= byte.MaxValue);
+            Contract.Assert(AmountOfBinsIsValid(amountOfBins), $"{nameof(amountOfBins)} should be in range [1, {MaxBins}] and be a power of 2.");
+            _locationsPerBin = machinesPerBin;
+            _entriesPerMachine = entriesPerMachine;
+            _bins = new MachineLocation[amountOfBins];
+        }
+
+        private static bool AmountOfBinsIsValid(int amount)
+        {
+            return amount > 0 && amount <= MaxBins &&
+                ((amount & (amount - 1)) == 0); // Is power of 2.
+        }
+
+        /// <nodoc />
+        public void AddLocation(MachineLocation location) => ProcessLocation(location, valueToSet: location);
+
+        /// <nodoc />
+        public void RemoveLocation(MachineLocation location) => ProcessLocation(location, valueToSet: default);
+
+        private void ProcessLocation(MachineLocation location, MachineLocation valueToSet)
+        {
+            var keyBytes = new byte[location.Data.Length + 1];
+            location.Data.CopyTo(keyBytes, 0);
+            for (byte entry = 0; entry < _entriesPerMachine; entry++)
+            {
+                keyBytes[keyBytes.Length - 1] = entry;
+                var key = _contentHasher.GetContentHash(keyBytes);
+                var index = BitConverter.ToUInt16(key.ToByteArray(), 1);
+                _bins[index % _bins.Length] = valueToSet;
+            }
+
+            // Invalidate current configuration.
+            _machinesForBins = null;
+        }
+
+        /// <nodoc />
+        public MachineLocation[] GetLocations(ContentHash hash)
+        {
+            _machinesForBins ??= ComputeBins();
+            var index = BitConverter.ToUInt16(hash.ToByteArray(), 1);
+            return _machinesForBins[index % _bins.Length];

```

Comment:
>_machinesForBins [](start = 19, length = 16)

Race condition here with _machineForBins being marked null. You should capture into a local or lock around mutations to _machinesForBins.